Fix chi^2 calculation with weights

jsdillon opened the pull request:
Added unittest to expose the issue.

This fixes #7

AaronParsons commented:
2 questions:
1) Should the weights be treated as 1/sigma or 1/sigma**2?  Inverse variance weighting would suggest the latter.

2) The _chisq method takes wgts as an argument (without a default value), but supports special case of a length zero dictionary.  Why does it have this special case?

jsdillon commented:
Ok @AaronParsons , fixed... along with a whole huge numpy type issue.

Turns out the problem was in 
```            try: x,y,val = p.sparse_form(term[-1], eqnum, prm_order, f.flatten(), re_im_split)
            except(AttributeError): # happens if f is a scalar
                x,y,val = p.sparse_form(term[-1], eqnum, prm_order, f, re_im_split)```
While you can't call `.flatten()` on python types, you can call it on `np.float32`, etc. and it gives back a length 1 list (which is a very confusing choice, IMHO). So that except was never getting triggered. I've rewritten the two cases.

Lmk if you approve of the PR.

jsdillon commented:
Also removed the `put_matrix` functions, which have apparently been deprecated.
